### PR TITLE
Fix domain erasure in realm rewriting

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR fixes an issue in authentication rewriting: previously `maybeRewriteRealmRef` did not remove domains, even when this was expected. This has very limited customer impact, since the enclosing `maybeRewriteForOlderVersion` method is called in conjunction with `Authentication.encode()` in most cases. `Authentication.encode()` calls `writeTo`, which in turn omits domains as necessary, based on a version check.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 1]
```diff
@@ -882,11 +882,12 @@ public static Authentication newApiKeyAuthentication(AuthenticationResult<User>
...
     }
 
-    private static RealmRef maybeRewriteRealmRef(Version streamVersion, RealmRef realmRef) {
+    // pkg-private for testing
+    static RealmRef maybeRewriteRealmRef(Version streamVersion, RealmRef realmRef) {
         if (realmRef != null && realmRef.getDomain() != null && streamVersion.before(VERSION_REALM_DOMAINS)) {
             logger.info("Rewriting realm [" + realmRef + "] without domain");
             // security domain erasure
-            new RealmRef(realmRef.getName(), realmRef.getType(), realmRef.getNodeName(), null);
+            return new RealmRef(realmRef.getName(), realmRef.getType(), realmRef.getNodeName(), null);
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:890]: Sneaky!

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java [hunk 2]
```diff
@@ -507,6 +508,73 @@ public void testBwcWithStoredAuthenticationHeaders() throws IOException {
         assertThat(authenticationV7.encode(), equalTo(headerV7));
     }
 
+    public void testMaybeRewriteForOlderVersionErasesDomainForVersionsBeforeDomains() {
+        final TransportVersion olderVersion = TransportVersionUtils.randomVersionBetween(
+            random(),
+            Authentication.VERSION_REALM_DOMAINS.calculateMinimumCompatVersion(),
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java:514]: Looks like this is a deprecated API, and I think it's easy to find the replacement.
Also, hard-coded versions, especially for the bounds of the random generator, are OK too.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java [hunk 3]
```diff
@@ -507,6 +508,73 @@ public void testBwcWithStoredAuthenticationHeaders() throws IOException {
...
+            random(),
+            Authentication.VERSION_REALM_DOMAINS,
+            TransportVersionUtils.getPreviousVersion()
+        );
+        final Authentication authentication = AuthenticationTestHelper.builder()
+            .realm() // randomize to test both when realm is null on the original auth and non-null, instead of setting `underDomain`
+            .transportVersion(TransportVersionUtils.randomVersionBetween(random(), olderVersion, null))
+            .build();
+
+        final Authentication actual = authentication.maybeRewriteForOlderVersion(olderVersion);
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java:541]: Maybe use a different version when rewriting from the original.

**Review by albertzaharovits — APPROVED** [review]
LGTM
Nice catch and thank you for the test.

**Comment by albertzaharovits** [comment]
I have labeled it as a `bug`.
For (bug-fix) PRs (not GH issues) the GH version labels reflect the ES versions that contain the fix, rather than the released ES versions affected by the bug.

Take care to pull in the changes pushed in by the bot in https://github.com/elastic/elasticsearch/pull/93276#issuecomment-1406176363 before you push to the PR branch.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java (hunk 1 above)
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:890]: Nice catch!

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java (hunk 2 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java:514]: Sounds good, I'll go with a hard-coded version -- unfortunately, there isn't a replacement method just a note stating "Code calling this should be refactored to not do this" 😅

**Comment by n1v0lg** [comment]
@albertzaharovits thanks for the review!

> I have labeled it as a bug.

👍 I went with labeling `non-issue` since there shouldn't be customer impact which might make the release note regarding a bug fix a bit of red herring for end-users. I don't feel strongly though, all good with the `bug` label.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java (hunk 2 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java:514]: Ah nice, there's a new util method `randomPreviousCompatibleVersion` which does the trick.

**Pull request merged**